Add bluetooth API doc

apoudret-microej opened the pull request.

gjadi-mej commented on ApplicationDeveloperGuide/bluetooth.rst:
```diff
@@ -0,0 +1,255 @@
+
+- Either device sends a pairing request or security request to the other device
+- Both devices share their I/O capabilities
+- If the I/O capabilities of the devices allow to create a connection with MITM , the Passkey Entry method is used:
+
+  - The device with display capability displays a generated passkey on its user interface
+  - The device with input capability reads the passkey from the user input and sends it to the device with display capability
+  - The device with display capability checks that the passkey match
+- Otherwise, the Just Works method is used and the pairing is complete. This method does not prevent MITM attacks.
```
This part on how to avoid MITM is confusing to me.
From the first sentence, my understanding is that, if the connection between the two devices is not secured (MITM is possible), then the `Passkey Entry` method is used. Otherwise (no MITM possible) the `Just Works` method is used.

However, the last sentence on `Just Works` method says this method does not prevent MITM, but that shouldn't be a concern because otherwise the `Passkey Entry` method would have been used.
What am I missing?

apoudret-microej replied:
you're right there is a mistake in the first sentence. Passkey Entry allows MITM **protection**

apoudret-microej replied:
fixed

gjadi-mej commented on ApplicationDeveloperGuide/networking.rst:
```diff
@@ -9,6 +9,7 @@ Networking
    
    networkingFoundationLibraries
    networkingAddonLibraries
+   bluetooth
```
it's a bit confusing.
In the app developer guide, bluetooth is under networking section.
In vee port guide, it has its own section.
From my POV, it's confusing to look for bluetooth under the networking section.
What do you think about moving it one section above? in the `libraries.rst` instead of `networking.rst`?

gjadi-mej commented on VEEPortingGuide/bluetooth.rst:
```diff
@@ -31,8 +31,12 @@ The Bluetooth Foundation Library provides a way to manage and configure Bluetoot
 Dependencies
 ============
 
```
In the `Use` section, could you link to the Bluetooth API page in the application developer guide?
This would make it easier to find it and avoid repeating the same thing (for example the `Use` section uses MMM dependency only instead of gradle + MMM.

apoudret-microej replied:
agree, will do like in https://docs.microej.com/en/latest/VEEPortingGuide/uiMicroui.html#use

apoudret-microej replied:
done

apoudret-microej commented on ApplicationDeveloperGuide/bluetooth.rst:
```diff
@@ -0,0 +1,255 @@
+Basic Knowledge and APIs
+------------------------
+
+.. _BluetoothAdapter: https://repository.microej.com/javadoc/microej_5.x/apis/ej/bluetooth/BluetoothAdapter.html
```
what do  you mean? the links are referenced later (e.g. in classes summary)

gjadi-mej replied:
Ohh my bad, I missremembered the comment syntax and I am used to having the links at the end of the document.

thomasponsmicroej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+
+A device can call the `BluetoothConnection.sendPairRequest()`_ API to initiate pairing.
+The `ConnectionListener.onPairRequest()`_ hook is called when the device receives a pairing request.
+It can call the `BluetoothConnection.sendPairResponse()`_ API to accept or deny the pairing.
```
if the device doesn't send any response, what is the consequence on the communication between the 2 devices?

apoudret-microej replied:
after a few seconds without a response, onPairCompleted() will be called with success=false

thomasponsmicroej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+A characteristic has property flags, which indicate to the other devices how the characteristic can be used (whether it can be written, whether it provides notifications, etc.).
+A characteristic may have descriptors, which allow to describe or configure the characteristic in a specific way.
+Every attribute (characteristic or descriptor) has permission flags, which control its access (read-only, read/write, requires authentication, etc.).
+Services and attributes are all identified by 16-bit UUID. If a service or attribute is standard, the relevant specification indicates its UUID.
```
```suggestion
Services and attributes are all identified by a 16-bit UUID. If a service or attribute is standard, the relevant specification indicates its UUID.
```

apoudret-microej replied:
fixed

thomasponsmicroej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+- A **read request** on a characteristic of a **discovered service** allows to **request data** from the device it belongs to.
+  The device which discovered the service sends a read request and the device which owns the service sends back a read response with the data.
+- A **write request** on a characteristic of a **discovered service** allows to **send data** to the device it belongs to.
+  The device which discovered the service sends a write request with the data and the device which owns the service sends back a read response.
```
```suggestion
  The device which discovered the service sends a write request with the data and the device which owns the service sends back a read response.
```
a write response?

apoudret-microej replied:
fixed

thomasponsmicroej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+  The device which owns the service sends a notification with the data to the desired device.
+  BLE provides a built-in way to subcribe to the notifications of a characteristic of a discovered service, by sending a write request on its CCC descriptor.
+
+For the read request procedure, a device can call the send `BluetoothConnection.sendReadRequest()`_ API to send a read request.
```
```suggestion
For the read request procedure, a device can call the `BluetoothConnection.sendReadRequest()`_ API to send a read request.
```

apoudret-microej replied:
fixed

thomasponsmicroej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+- `BluetoothConnection`_: Performs operations related to a specific device connection (disconnect, pair, discover remote GATT services, send GATT requests)
+- `BluetoothService`_: Represents a local or remote GATT service
+- `ConnectionListener`_ and `DefaultConnectionListener`_: Callbacks for all events not related to a specific GATT service
+- `LocalServiceListener`_ and `DefaultLocalServiceListener`_: Callbacks for events related to a specific local GATT service
```
local = service that the device own and remote = service that the device discovered?
if yes, do you think it is worth mentioning?

thomasponsmicroej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+
+For the notification procedure, a device can call the send `BluetoothConnection.sendNotification()`_ API to send a notification with the data.
+The `RemoteServiceListener.onNotificationReceived()`_ hook is called with the data when a device receives a notification.
+
```
on this part I feel that a little sequence diagram between the two devices could help understand the request and callbacks. 
Maybe the same would apply for the discovery/pairing part in the section before.

apoudret-microej replied:
i added some diagrams. we can still add some more later

thomasponsmicroej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+BLE devices use characteristics to transfer data. There are 3 main procedures:
+
+- A **read request** on a characteristic of a **discovered service** allows to **request data** from the device it belongs to.
+  The device which discovered the service sends a read request and the device which owns the service sends back a read response with the data.
```
is the read/write response mandatory (what if the device fails to answer?)

apoudret-microej replied:
after a few seconds without a response, the app should receive a hook (onReadCompleted() with error status or onDisconnected()?)

apoudret-microej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+
+BLE introduces two roles of devices: the Central and Peripheral roles.
+A Central device scans nearby Peripheral devices and initiates the connection, while a Peripheral device advertises (broadcasts) and listens for connection requests.
+In this regard, a Peripheral device can be thought of as a Wi-Fi Access Point while a Peripheral device can be thought of as a Wi-Fi Station.
```
fixed

apoudret-microej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+
+- Either device sends a pairing request or security request to the other device
+- Both devices share their I/O capabilities
+- If the I/O capabilities of the devices allow to create a connection with MITM protection, the Passkey Entry method is used:
```
i replaced MITM by MITM attacks, do you think it is sufficient?

thomasponsmicroej replied:
👍

apoudret-microej commented on ApplicationDeveloperGuide/bluetoothAPI.rst:
```diff
@@ -0,0 +1,209 @@
+  - The device with display capability displays a generated passkey on its user interface
+  - The device with input capability reads the passkey from the user input and sends it to the device with display capability
+  - The device with display capability checks that the passkey match
+- Otherwise, the Just Works method is used and the pairing is complete. This method does not prevent MITM attacks.
```
fixed

apoudret-microej commented:
> Other topics that could be added (now or later):
> 
> * a section about the write requests/response mechanism for large files (> MTU size)
> * how to design the app for maximum throughput

i added the Write Without Response procedure and i added a new section with guidelines on how to achieve maximum throughput
